File: web/src/api/scheduledTask.ts
/**
 * 定时任务 API
 */

import { get, post, del } from '@/utils/request'

/** 与后端一致：execute=普通函数；table_*=表格增改删（走 /_callback） */
export type ScheduledTaskAction =
  | 'execute'
  | 'table_create'
  | 'table_update'
  | 'table_delete'

export interface CreateScheduledTaskReq {
  name: string
  full_code_path: string
  action?: ScheduledTaskAction
  method?: string
  payload?: Record<string, unknown> | string
  request_user?: string
  request_user_dept?: string
  schedule_type: 'atime' | 'cron' | 'every'
  run_at?: string
  cron_expr?: string
  interval_seconds?: number
  max_runs?: number
  timezone?: string
}

export interface ScheduledTaskItem {
  id: number
  name: string
  user: string
  app: string
  full_code_path: string
  action?: string
  method?: string
  payload?: string
  request_user?: string
  request_user_dept?: string
  created_by: string
  schedule_type: string
  run_at: string
  next_run_at?: string
  cron_expr?: string
  interval_seconds?: number
  max_runs?: number
  timezone?: string
  status: string
  run_count: number
  error_message?: string
  created_at: string
}

export interface ScheduledTaskExecutionItem {
  id: number
  task_id: number
  executed_at: string
  status: string
  duration_millis?: number
  request_payload: string
  response_payload: string
  error_message?: string
  trace_id?: string
  created_at: string
}

export interface ListScheduledTasksResp {
  list: ScheduledTaskItem[]
  total: number
}

export interface ListScheduledTaskExecutionsResp {
  list: ScheduledTaskExecutionItem[]
  total: number
}

export function createScheduledTask(data: CreateScheduledTaskReq): Promise<ScheduledTaskItem> {
  // 直接传对象，后端存成 json.RawMessage 即 {"a":1} 的字节；不要 JSON.stringify(payload) 否则会变成字符串导致执行时 unmarshal 报错
  const payload = data.payload ?? {}
  return post<ScheduledTaskItem>('/workspace/api/v1/scheduled_tasks', {
    name: data.name,
    full_code_path: data.full_code_path,
    action: data.action ?? 'execute',
    method: data.method || 'POST',
    payload,
    request_user: data.request_user,
    request_user_dept: data.request_user_dept,
    schedule_type: data.schedule_type,
    run_at: data.run_at,
    cron_expr: data.cron_expr,
    interval_seconds: data.interval_seconds,
    max_runs: data.max_runs ?? 0,
    timezone: data.timezone
  })
}

/** full_code_path 为前缀：返回该路径及子路径下的任务（目录节点可看到子表单的定时任务） */
export function listScheduledTasks(params: {
  full_code_path?: string
  status?: string
  page?: number
  page_size?: number
}): Promise<ListScheduledTasksResp> {
  return get<ListScheduledTasksResp>('/workspace/api/v1/scheduled_tasks', params)
}

export function cancelScheduledTask(id: number): Promise<void> {
  return del<void>(`/workspace/api/v1/scheduled_tasks/${id}`)
}

export function listScheduledTaskExecutions(
  taskId: number,
  params?: { status?: string; page?: number; page_size?: number }
): Promise<ListScheduledTaskExecutionsResp> {
  return get<ListScheduledTaskExecutionsResp>(
    `/workspace/api/v1/scheduled_tasks/${taskId}/executions`,
    params
  )
}

File: dto/scheduled_task.go
package dto

import "encoding/json"

// CreateScheduledTaskReq 创建定时任务请求
type CreateScheduledTaskReq struct {
	Name            string          `json:"name" binding:"required"`           // 任务名称
	FullCodePath    string          `json:"full_code_path" binding:"required"` // 如 /system/official/message/send_message_to_users.form
	Action          string          `json:"action"`                            // 执行动作：空/execute/form(普通函数)、table_create、table_update、table_delete
	Method          string          `json:"method"`                            // GET/POST，默认 POST
	Payload         json.RawMessage `json:"payload"`                           // 请求体 JSON
	RequestUser     string          `json:"request_user"`                      // 以谁执行，空则用当前用户
	RequestUserDept string          `json:"request_user_dept"`                 // 请求用户部门路径（透传，空则用创建人部门）
	ScheduleType    string          `json:"schedule_type" binding:"required"`  // atime / cron / every
	RunAt           string          `json:"run_at"`                            // 仅 schedule_type=atime 时必填；cron/every 由服务端按创建时间自动补
	CronExpr        string          `json:"cron_expr"`                         // schedule_type=cron 时必填
	IntervalSeconds int64           `json:"interval_seconds"`                  // schedule_type=every 时必填
	MaxRuns         int             `json:"max_runs"`                          // schedule_type=every 时可选，0=不限制
	Timezone        string          `json:"timezone"`                          // 可选
}

// ScheduledTaskItem 定时任务列表项
type ScheduledTaskItem struct {
	ID              int64   `json:"id"`
	Name            string  `json:"name"`
	User            string  `json:"user"`
	App             string  `json:"app"`
	FullCodePath    string  `json:"full_code_path"`
	Action          string  `json:"action"`
	Method          string  `json:"method"`
	Payload         string  `json:"payload"`
	RequestUser     string  `json:"request_user"`
	RequestUserDept string  `json:"request_user_dept"`
	CreatedBy       string  `json:"created_by"`
	ScheduleType    string  `json:"schedule_type"`
	RunAt           string  `json:"run_at"` // atime 的执行时间；cron/every 的生效时间
	NextRunAt       *string `json:"next_run_at,omitempty"`
	CronExpr        string  `json:"cron_expr,omitempty"`
	IntervalSeconds int64   `json:"interval_seconds,omitempty"`
	MaxRuns         int     `json:"max_runs,omitempty"`
	Timezone        string  `json:"timezone,omitempty"`
	Status          string  `json:"status"`
	RunCount        int     `json:"run_count"`
	ErrorMessage    string  `json:"error_message,omitempty"`
	CreatedAt       string  `json:"created_at"`
}

// ScheduledTaskExecutionItem 执行记录列表项
type ScheduledTaskExecutionItem struct {
	ID              int64  `json:"id"`
	TaskID          int64  `json:"task_id"`
	ExecutedAt      string `json:"executed_at"`
	Status          string `json:"status"`
	DurationMillis  int64  `json:"duration_millis,omitempty"`
	RequestPayload  string `json:"request_payload"`  // JSON 字符串
	ResponsePayload string `json:"response_payload"` // JSON 字符串
	ErrorMessage    string `json:"error_message,omitempty"`
	TraceID         string `json:"trace_id,omitempty"`
	CreatedAt       string `json:"created_at"`
}

// ListScheduledTasksResp 定时任务列表响应
type ListScheduledTasksResp struct {
	List  []ScheduledTaskItem `json:"list"`
	Total int64               `json:"total"`
}

// ListScheduledTaskExecutionsResp 定时任务执行记录列表响应
type ListScheduledTaskExecutionsResp struct {
	List  []ScheduledTaskExecutionItem `json:"list"`
	Total int64                        `json:"total"`
}
